### ISSUE TITLE
Doctests now run on non-exported `macro_rules!`, but can't actually test the macro

### ISSUE THREAD

**Issue by cuviper** [rust-lang/rust#97030]
Since #88038 was resolved with #96630, doctests now run on `macro_rules!` even if the macro is not exported.

This could be considered a breaking change -- it did break CI for rayon when [this macro example](https://github.com/rayon-rs/rayon/blob/9dc52d76b12f3da13b87dd509ff29a801177b8d3/src/delegate.rs#L14-L19) for internal use suddenly became a real test and failed.

The problem is, I can't figure out how to make a working doctest for that, since it's not exported in the public API. What is the intended way to do this? And if there's not really a way to do it, then maybe this change should be reverted so there isn't any break.

I thought of `#[cfg_attr(doctest, macro_export)]`, which does make even older Rust try to test it, but that still can't reach the macro because it's still using the non-`doctest` build of the crate.

---

PS: I ended up making unit tests for that rayon macro, which is probably better in that case, but that doesn't answer how non-exported macro tests should work.

**Comment by GuillaumeGomez** [comment]
cc @m-ysk

**Comment by Nemo157** [comment]
One reason to keep the change is that doctests on private functions also run, even though they similarly can't access the function they are documenting. For consistency it makes sense that all doctests just run, and it's up to the dev to either ` ```ignore ` or to use `//` comment syntax to make it not a doctest.

**Comment by jyn514** [comment]
I don't think this is a bug. This is consistent with tests for other non-macro items, which also can't use themselves in the test. The test might be testing something other than the defined item, in which case it's still useful to run.

**Comment by jonhoo** [comment]
Carrying over from https://github.com/rust-lang/rust/issues/98735, this should be marked as `regression-from-stable-to-stable`. If it is indeed intended breakage, it should be added to the compatibility notes for the 1.62.0 release so that those adopting the new release understand why things broke and how to fix it.

**Comment by CAD97** [comment]
PR to add it to a rustdoc section: https://github.com/rust-lang/rust/pull/98749
